services/addons/images/firmware_manager/firmware_manager.py

```diff
@@ -259,18 +299,66 @@ def check_for_upgrades():
             ):
                 continue
 
+            # Check if latest ping was unsuccessful
+            if not was_latest_ping_successful_for_rsu(rsu["ipv4_address"]):
+                logging.info(
+                    f"Skipping firmware upgrade for '{rsu['ipv4_address']}': device is unreachable"
+                )
+                continue
+
             # Add the RSU to the upgrade queue and record the necessary upgrade information
             logging.info(
                 f"Adding '{rsu["ipv4_address"]}' to the firmware manager upgrade queue"
             )
             upgrade_queue.extend([rsu["ipv4_address"]])
             upgrade_queue_info[rsu["ipv4_address"]] = rsu
-            logging.info(f"Firmware upgrade successfully started for '{rsu["ipv4_address"]}'")
+            logging.info(
+                f"Firmware upgrade successfully started for '{rsu["ipv4_address"]}'"
+            )
 
         # Start any processes that can be started
         start_tasks_from_queue()
 
 
+def was_latest_ping_successful_for_rsu(rsu_ip):
+    query = f"select result from ping where rsu_id=(select rsu_id from rsus where ipv4_address='{rsu_ip}') order by timestamp desc limit 1"
+    query_result = pgquery.query_db(query)
+    if len(query_result) == 0 or len(query_result[0]) == 0:
+        # no ping results have been recorded for this RSU
+        return False
+    latest_ping_successful = query_result[0][0]
+    logging.info(f"Latest ping result for '{rsu_ip}': {latest_ping_successful}")
+    return latest_ping_successful
+
+
+def increment_consecutive_failure_count_for_rsu(rsu_ip):
+    upsert_query = f"insert into consecutive_firmware_upgrade_failures (rsu_id, consecutive_failures) values ((select rsu_id from rsus where ipv4_address='{rsu_ip}'), 1) on conflict (rsu_id) do update set consecutive_failures=consecutive_firmware_upgrade_failures.consecutive_failures+1"
+    pgquery.write_db(upsert_query)
+
+
+def reset_consecutive_failure_count_for_rsu(rsu_ip):
+    upsert_query = f"insert into consecutive_firmware_upgrade_failures (rsu_id, consecutive_failures) values ((select rsu_id from rsus where ipv4_address='{rsu_ip}'), 0) on conflict (rsu_id) do update set consecutive_failures=0"
+    pgquery.write_db(upsert_query)
+
+
+def is_rsu_at_max_retries_limit(rsu_ip):
+    max_retries = int(os.environ.get("FW_UPGRADE_MAX_RETRY_LIMIT", "3"))
+    query_result = pgquery.query_db(
+        f"select consecutive_failures from consecutive_firmware_upgrade_failures where rsu_id=(select rsu_id from rsus where ipv4_address='{rsu_ip}')"
+    )
+    if len(query_result) == 0 or len(query_result[0]) == 0:
+        # no failures have been recorded for this RSU, so it cannot be at the limit
+        return False
+    consecutive_failures = query_result[0][0]
+    return consecutive_failures >= max_retries
+
+
+def log_max_retries_reached_incident_for_rsu_to_postgres(rsu_ip):

```

Comment:
Do you think it'd make sense to also log the attempted firmare version that was causing the issue? It's possible we'd see repeated failures from various versions and might be nice to have that version in here to know exactly which one failed

Reply:
Yeah I think this makes sense. I've modified the max retry limit reached instance tracking to include the firmware id corresponding to the target firmware version at the time of upgrading.